### PR TITLE
OOP #1

### PR DESCRIPTION
**Pull request by TalNovatzky**
Authors: @shadston,  @TalNovatzky

### CODE REVIEW

File: shm/Cargo.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+#include <string>
+
+#include<string>
```
Comment (WBigaj) [inline shm/Cargo.h]: Why include string twice in the same file?

File: shm/Island.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+#include "Coordinates.h"
```
Comment (WBigaj) [inline shm/Island.h:3]: I think there should be a newline between pragma and include.

File: shm/Ship.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    Ship(int capacity, int maxCrew, int speed, const std::string& name, size_t id)
+        : capacity_(capacity), maxCrew_(maxCrew), crew_(0), speed_(speed), name_(name), id_(id) {}
+    Ship(int maxCrew, int speed, size_t id)
+        : Ship(0, maxCrew, speed, "", id) {}
+
+    void setName(const std::string& name) { name_ = name; }
+
+    Ship& operator-=(size_t num) {
+        crew_ -= num;
+        return *this;
```
Comment (WBigaj) [inline shm/Ship.cpp]: How are you handling negative size of the crew?

File: shm/Ship.cpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    }
+    Ship& operator+=(size_t num) {
+        crew_ += num;
+        return *this;
```
Comment (WBigaj) [inline shm/Ship.cpp]: How are checking if you have space in the ship you crewman you want to add?

**Review by WBigaj — COMMENTED** [review]
I have checked your code and gave my feedback.

File: shm/Cargo.h (hunk 1 above)
Comment (nauka-programowania-MA) [reply shm/Cargo.h]: +1

File: shm/Player.cpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    return ship_.getSpeed();
+}
+
+Cargo* getCargo(size_t index) const{
+    return _ship->getCargo(index).get();
+}
+
+void Player::checkAvailableSpace()
+{
+    size_t usedSpace = std::accumulate(_ship->getCargos().begin(), _ship->getCargos().end(), 0, [](size_t total, auto item) { return total + item->getAmount(); });
```
Comment (nauka-programowania-MA) [inline shm/Player.cpp]: too long line

File: shm/Ship.cpp (hunk 3 above)
Comment (nauka-programowania-MA) [reply shm/Ship.cpp]: +1

File: shm/Ship.cpp (hunk 4 above)
Comment (nauka-programowania-MA) [reply shm/Ship.cpp]: +1

File: shm/Ship.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#include <string>
+
+class Ship {
+public:
```
Comment (nauka-programowania-MA) [inline shm/Ship.cpp]: Why you have this in .cpp file?

File: shm/Ship.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
+#include <string>
+
+class Ship {
+public:
+    Ship() {}
+    Ship(int capacity, int maxCrew, int speed, const std::string& name, size_t id)
+        : capacity_(capacity), maxCrew_(maxCrew), speed_(speed), name_(name), id_(id) {}
+    Ship(int maxCrew, int speed, size_t id)
+        : Ship(0, maxCrew, speed, "", id) {}
```
Comment (nauka-programowania-MA) [inline shm/Ship.h]: move to .cpp

File: shm/Ship.cpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    size_t getId() const { return id_; }
+
+private:
+    size_t capacity_;
+    size_t maxCrew_;
+    size_t crew_;
+    size_t speed_;
+    std::string name_;
+    const size_t id_;
+};
```
Comment (nauka-programowania-MA) [inline shm/Ship.cpp]: Something really wrong here. You should include "Ship.h" here and create only definition.

**Review by nauka-programowania-MA — CHANGES_REQUESTED** [review]

File: shm/Cargo.cpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
+#include "Cargo.h"
+
+	Cargo::Cargo (std::string name, size_t amount, size_t basePrice) :
```
Comment (kuchichan) [inline shm/Cargo.cpp]: Good idea will be to add .clang-format file, so all co-owners (workers ;p) will use the same formatting.

File: shm/Cargo.cpp [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+				basePrice_(basePrice)
+				{}
+
+	Cargo::Cargo& operator+=(size_t amount) {
+		amount_ += amount;
+		return *this;
+	}
+
+	Cargo::Cargo& operator-=(size_t amount) {
+		amount_ -= amount;
```
Comment (kuchichan) [inline shm/Cargo.cpp]: Check for negative numbers

File: shm/Map.cpp [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <algorithm>
+#include <random>
+
+#include "Island.h"
+
+Map::Map() {
+    for (int i = 0; i < islandsCount; i++) {
+        std::random_device rd;
+        std::mt19937 gen(rd());
```
Comment (kuchichan) [inline shm/Map.cpp]: I think this might be instantiated outside the loop

File: shm/Cargo.h (hunk 1 above)
Comment (shadston) [reply shm/Cargo.h]: fixed

File: shm/Island.h (hunk 2 above)
Comment (shadston) [reply shm/Island.h:3]: why? is that good practice?

File: shm/Map.cpp (hunk 11 above)
Comment (shadston) [reply shm/Map.cpp]: fixed

File: shm/Cargo.cpp (hunk 10 above)
Comment (shadston) [reply shm/Cargo.cpp]: fixed

File: shm/Ship.cpp (hunk 3 above)
Comment (shadston) [reply shm/Ship.cpp]: fixed

File: shm/Ship.cpp (hunk 4 above)
Comment (shadston) [reply shm/Ship.cpp]: fixed

File: shm/Player.cpp (hunk 5 above)
Comment (shadston) [reply shm/Player.cpp]: fixed

File: shm/Cargo.cpp (hunk 9 above)
Comment (shadston) [reply shm/Cargo.cpp]: you are right, its a good idea :D

File: shm/Ship.cpp (hunk 8 above)
Comment (shadston) [reply shm/Ship.cpp]: yeee, I guess working late is wrong ;p should be fixed ;)

File: shm/Ship.h (hunk 7 above)
Comment (shadston) [reply shm/Ship.h]: moved

File: shm/Ship.cpp (hunk 6 above)
Comment (shadston) [reply shm/Ship.cpp]: In the begining I didn't know to properly split things between c++ and header file, but I belive I do know now ;p

File: shm/Island.cpp [hunk 12]
```diff
@@ -0,0 +1,6 @@
+#include "Island.h"
+
+Coordinates Island::getCoordinates() const
+{
+	{ return position_; }
```
Comment (nauka-programowania-MA) [inline shm/Island.cpp:5]: {} redundant here

File: shm/Player.cpp [hunk 13]
```diff
@@ -0,0 +1,36 @@
+#include "Player.h"
+
+#include <numeric>
+
+Player::Player(Ship ship, size_t money)
+    : ship_(std::make_unique<Ship>(ship)), money_(money) {}
```
Comment (nauka-programowania-MA) [inline shm/Player.cpp:6]: I prefer pass std::unique_ptr as argument to player.

File: shm/Player.cpp [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    return ship_.get();
+}
+
+size_t Player::getSpeed() const {
+    return ship_->getSpeed();
+}
+
+Cargo* Player::getCargo(size_t index) const {
+    if (ship_) {
+        return ship_->getCargo(index).get();
```
Comment (nauka-programowania-MA) [inline shm/Player.cpp:26]: What if index is out of range? You got nullptr? If yes, you will call get() on uninitialized unique_ptr and got crash.

File: shm/Ship.cpp [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+Ship& Ship::operator+=(size_t num) {
+    crew_ += num;
+    if (crew_ > maxCrew_) {
+        crew_ = maxCrew_;
+    }
+    return *this;
+}
+
+std::shared_ptr<Cargo> Ship::getCargo(size_t index) {
+    return cargo_[index];
```
Comment (nauka-programowania-MA) [inline shm/Ship.cpp:31]: Are you validate somehow index?

**Review by nauka-programowania-MA — APPROVED** [review]

**Comment by nauka-programowania-MA** [comment]
{
"requirement_ids": [2568, 2568, 2568, 2568, 2568, 2568, 2568, 2568, 2569, 2570],
"points": [5, 5, 5, 5, 5, 5, 5, 5, 20, 6],
"users": ["shadston", "TalNovatzky"]
}

**Pull request closed without merging**